File: geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-api/src/main/java/org/locationtech/geomesa/fs/storage/api/FileSystemStorageFactory.java
/***********************************************************************
 * Copyright (c) 2013-2019 Commonwealth Computer Research, Inc.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Apache License, Version 2.0
 * which accompanies this distribution and is available at
 * http://www.opensource.org/licenses/apache2.0.php.
 ***********************************************************************/

package org.locationtech.geomesa.fs.storage.api;

import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.fs.FileContext;
import org.apache.hadoop.fs.Path;
import org.opengis.feature.simple.SimpleFeatureType;

import java.util.Optional;

public interface FileSystemStorageFactory {

    /**
     * The file encoding used by this factory
     *
     * @return encoding
     */
    String getEncoding();

    /**
     * Attempt to load a storage from the given root path. If there is no compatible storage at the path,
     * will return nothing.
     *
     * @param fc file context
     * @param conf configuration
     * @param root storage root path
     * @return storage, if path is compatible with this factory
     */
    Optional<FileSystemStorage> load(FileContext fc, Configuration conf, Path root);

    /**
     * Create a new storage instance at the given root path, which should be empty
     *
     * @param fc file context
     * @param conf configuration
     * @param root storage root path
     * @param sft simple feature type to store
     * @return the newly created storage
     */
    FileSystemStorage create(FileContext fc, Configuration conf, Path root, SimpleFeatureType sft);
}

File: geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-api/src/main/scala/org/locationtech/geomesa/fs/storage/api/FileSystemStorageFactory.scala
/***********************************************************************
 * Copyright (c) 2013-2025 General Atomics Integrated Intelligence, Inc.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Apache License, Version 2.0
 * which accompanies this distribution and is available at
 * https://www.apache.org/licenses/LICENSE-2.0
 ***********************************************************************/

package org.locationtech.geomesa.fs.storage.api

import java.util.ServiceLoader

/**
  * Factory for creating and loading file system storage implementations
  */
trait FileSystemStorageFactory {

  /**
    * The file encoding used by this factory
    *
    * @return encoding
    */
  def encoding: String

  /**
    * Create a new storage instance pointing at the given context path
    *
    * @param context file context
    * @param metadata metadata persistence
    * @return
    */
  def apply(context: FileSystemContext, metadata: StorageMetadata): FileSystemStorage
}

object FileSystemStorageFactory {

  import scala.collection.JavaConverters._

  lazy val factories: Seq[FileSystemStorageFactory] = ServiceLoader.load(classOf[FileSystemStorageFactory]).asScala.toSeq

  /**
   * Load a factory with the given encoding
   *
   * @param encoding file encoding
   * @return
   */
  def apply(encoding: String): FileSystemStorageFactory =
    factories.find(_.encoding.equalsIgnoreCase(encoding)).getOrElse {
      throw new IllegalArgumentException(s"Could not find a factory class for encoding '$encoding'. " +
        s"Factories are available for: ${factories.map(_.encoding).mkString(", ")}")
    }
}
